### PR TITLE
Don't error on missing units

### PR DESCRIPTION
**Pull request by hasufell**
Test units may be missing from plan.json, but be present
in the .cabal files.

----

Although you could just enable tests for your package, this becomes a bit difficult when making heavy use of `optional-packages: ...`, where some tests may not be desired to be built at all (e.g. because there's no valid plan).

### CODE REVIEW

File: cabal-hie/src/CabalHie/Main.hs [hunk 1]
```diff
@@ -110,12 +132,13 @@ generateHie tracer opts = do
         -- next we collect (directory, selector) pairs
         let componentPaths :: (Semigroup a, CL.HasBuildInfo a) => String -> C.CondTree C.ConfVar [d] a -> Peu r [(Path Absolute, Text)]
             componentPaths selector comp0  = do
-                (unit, _cn, _ci) <- maybe (die tracer $ "Cannot find unit for " ++ selector) return $
+                mcomp <- maybe (putWarning tracer WCan'tFindUnit ("Cannot find unit for " ++ selector) $> Nothing) (return . Just) $
```
Comment (phadej) [inline cabal-hie/src/CabalHie/Main.hs]: I'd fuse this with the `case` we are forced to do anyway later.

**Review by phadej — COMMENTED** [review]
Goog good.

File: cabal-hie/src/CabalHie/Main.hs [hunk 2]
```diff
@@ -109,15 +131,11 @@ generateHie tracer opts = do
...
-                (unit, _cn, _ci) <- maybe (die tracer $ "Cannot find unit for " ++ selector) return $
-                    Map.lookup (T.pack selector) components
-
-                let (_, comp) = simplifyCondTree ghcInfo (Map.mapKeys toCabal $ Plan.uFlags unit) comp0
-
-                return $ map (\dir -> (absDir dir, T.pack selector)) (L.toListOf (CL.hsSourceDirs . traverse) comp)
-
-        -- libraries
+            componentPaths selector comp0  =
+                maybe (putWarning tracer WCan'tFindUnit ("Cannot find unit for " ++ selector) $> [])
```
Comment (phadej) [inline cabal-hie/src/CabalHie/Main.hs]: Oh, I would prefered a `case`. This `maybe is unreadable. Sorry for not being explicit. I can do that change myself too later.

**Comment by phadej** [comment]
Thanks!

File: cabal-hie/src/CabalHie/Main.hs [hunk 3]
```diff
@@ -109,15 +130,15 @@ generateHie tracer opts = do
...
+            componentPaths selector comp0  =
+                case Map.lookup (T.pack selector) components of
+                    Nothing -> do
+                        putWarning tracer WCan'tFindUnit ("Cannot find unit for " ++ selector)
+                        pure []
+                    Just (unit, _cn, _ci) ->
+                        let (_, comp) = simplifyCondTree ghcInfo (Map.mapKeys toCabal $ Plan.uFlags unit) comp0
+                        in return $ map (\dir -> (absDir dir, T.pack selector)) (L.toListOf (CL.hsSourceDirs . traverse) comp)
 
-        -- libraries
```
Comment (hasufell) [inline cabal-hie/src/CabalHie/Main.hs:120]: oops... this was unintended...

Comment (phadej) [reply cabal-hie/src/CabalHie/Main.hs:120]: It's fine

**Pull request merged**